File: graph/edmonds-karp.rs
// loj-101.rs

use std::collections::VecDeque;

#[derive(Clone, Debug)]
struct Edge {
    to: usize,
    cap: i64,
    flow: i64,
    rev: (usize, usize)
}

struct Net(Vec<Vec<Edge>>);

impl Net {
    fn new(n: usize) -> Self {
        Self(vec![vec![];n+1])
    }
    fn add_edge(&mut self, from: usize, to: usize, cap: i64) {
        let to_edge = Edge {to, cap, flow: 0, rev: (to, self.0[to].len())};
        self.0[from].push(to_edge);
        let from_edge = Edge {to: from, cap: 0, flow: 0, rev: (from, self.0[from].len() - 1)};
        self.0[to].push(from_edge);
    }
    fn max_flow(&mut self, s: usize, t: usize) -> i64 {
        let n = self.0.len();
        let mut pa = vec![(0, 0);n+1];
        let mut res = 0;
        loop {
            let mut pf = vec![0;n+1];
            let mut dq = VecDeque::new();
            dq.push_back(s);
            pf[s] = i64::MAX;
            while let Some(v) = dq.pop_front() {
                for (i, ne) in self.0[v].iter().enumerate() {
                    if pf[ne.to] == 0 && ne.cap > ne.flow {
                        pf[ne.to] = pf[v].min(ne.cap - ne.flow);
                        pa[ne.to] = (v, i);
                        dq.push_back(ne.to);
                    }
                }
                if pf[t] != 0 {
                    break;
                }
            }
            if pf[t] == 0 {
                break;
            }
            let mut p = t;
            while pa[p].0 != 0 {
                let (x, y) = pa[p];
                self.0[x][y].flow += pf[t];
                let (z, w) = self.0[x][y].rev;
                self.0[z][w].flow -= pf[t];
                p = x;
            }
            res += pf[t];
        }
        res
    }
}

fn main() {
    let mut buf = String::new();
    std::io::stdin().read_line(&mut buf).ok();
    let tmp: Vec<usize> = buf.trim().split(' ').map(|x|x.parse::<usize>().unwrap()).collect::<Vec<usize>>();
    let (n, m, s, t) = (tmp[0], tmp[1], tmp[2], tmp[3]);
    let mut net = Net::new(n);
    for _ in 0..m {
        buf.clear();
        std::io::stdin().read_line(&mut buf).ok();
        let tmp: Vec<i64> = buf.trim().split(' ').map(|x|x.parse::<i64>().unwrap()).collect::<Vec<i64>>();
        net.add_edge(tmp[0] as usize, tmp[1] as usize, tmp[2]);
    }
    println!("{}", net.max_flow(s, t));
}

File: graph/edmonds-karp.cc
struct edmonds_karp {
    struct edge {
        int to;
        ll cap;
        ll flow;
        pair<int, int> rev;
        int mark;
    };

    vector<vector<edge>> edges;

    edmonds_karp(int n) : edges(n + 1) {}

    void add_edge(int from, int to, ll cap, int mark = false) {
        edges[from].push_back({to, cap, 0, make_pair(to, int(edges[to].size())), mark});
        edges[to].push_back({from, 0, 0, make_pair(from, int(edges[from].size() - 1)), mark});
    }

    ll run(int s, int t) {
        int n = edges.size();
        vector<pii> pa_(n + 1);
        ll res = 0;
        while (1) {
            vector<ll> pf(n + 1);
            deque<int> dq;
            dq.emplace_back(s);
            pf[s] = LLONG_MAX;
            while (dq.size()) {
                int v = dq.front(); dq.pop_front();
                for (auto&& [i, ne] : enumerate(edges[v])) {
                    if (pf[ne.to] == 0 and ne.cap > ne.flow) {
                        pf[ne.to] = min(pf[v], ne.cap - ne.flow);
                        pa_[ne.to] = {v, i};
                        dq.emplace_back(ne.to);
                    }
                }
                if (pf[t] != 0) {
                    break;
                }
            }
            if (pf[t] == 0) {
                break;
            }
            int p = t;
            while (pa_[p].first != 0) {
                auto [x, y] = pa_[p];
                edges[x][y].flow += pf[t];
                auto [z, w] = edges[x][y].rev;
                edges[z][w].flow -= pf[t];
                p = x;
            }
            res += pf[t];
        }
        return res;
    }
};
